### PR TITLE
Make HOP alias and mutation support opt-in

### LINKED ISSUE #149989: [dynamo][higher order ops] Make support_aliasing and support_input_mutation default to False
### 🐛 Describe the bug

This PR https://github.com/pytorch/pytorch/pull/148953/files adds support for checking for input mutation and aliasing for HOPs. Currently, the default is that we expect all HOPs to support input mutation and aliasing. And then we set it to False just for invoke_subgraph.

But we should do other way around, where we default the support to False. And then we just set it to True for a few HOPS that do support mutation/aliasing. This issue is to track this work.

### Error logs

_No response_

### Versions

NA

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185738

Higher order op subgraph tracing defaulted `supports_input_mutation` and `supports_aliasing` to `True`. Every caller was therefore permissive unless it explicitly opted out, and wrapping HOP subclasses also inherited permissive class attributes. HOPs that do not implement mutation or aliasing semantics could silently capture invalid body graphs.

This PR flips both speculation helper defaults to `False` and declares the same strict policy on `TorchHigherOrderOperatorVariable`. Concrete transparent wrappers that preserve body semantics opt in explicitly: `wrap`, `flex_gemm`, checkpoint, `dynamo_bypassing_wrapper`, set-grad/autocast wrappers, and `autograd.Function`. Strict wrapping HOPs inherit the safe defaults.

Flex-attention mask graphs need a narrower exception: read-only lifted captures may be views sharing storage, while outputs must not alias inputs or each other. A separate `supports_input_input_aliasing` capability is enabled only for `mask_fn` graphs in forward and direct backward tracing. Score-mod graphs retain broader output-alias support. Input mutation and mask callbacks returning captured tensor aliases are now rejected. Under `fullgraph=True` these violations are errors; under ordinary non-fullgraph compilation Dynamo records a graph break and falls back to eager dense attention, which can materialize the full scores matrix.

`Tensor.register_hook` permits returning the incoming gradient unchanged but forbids modifying it in place, so hook tracing explicitly enables aliasing while rejecting mutation and now reports `register_hook` in diagnostics. The sourceless `register_post_accumulate_grad_hook` path previously lowered through the ordinary hook HOP; without compiled autograd it now graph-breaks to preserve correct eager post-accumulate semantics instead of silently miscompiling.

The alias/mutation checker capabilities are keyword-only to avoid misordering adjacent booleans. Broadly enabling aliasing for `mask_fn` would admit unsupported output aliases, while cloning captured views would change semantics and add cost.

Fixes #149989
Generated by my agent

Benchmark Results:
- Old `True` defaults over 20 strict-mode compilations: median 27.430 ms, mean 27.510 ms.
- New `False` defaults: median 27.260 ms, mean 27.403 ms.

Test Plan:
- Reproduced and fixed the three Dr.CI failures in the compiled-autograd clone of `test_flex_attention_backward_mask_mod_input_aliasing_supported`.
- Ran focused forward/backward flex-attention tests and all dynamic-shape clones.
- Ran focused strict-mode, hints-wrapper, wrap, hook, BaseHOP, invoke-subgraph, `autograd.Function`, checkpoint, and export tests.
- Re-ran the prior CUDA invoke-subgraph reproducer.
- `git diff --check`.
- `lintrunner -a`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the new Dynamo dynamic-shapes CI failure by changing the wrap opt-in tests to use scalar tensors, which still exercises input mutation and output aliasing without introducing unrelated symbolic shape freevars in the dynamic-shapes clone. I also made strict_mode pass its strict aliasing/mutation flags explicitly. Targeted dynamic-shapes, HOP, hooks, base_hop, and export tests pass locally; lintrunner -a still fails on an unrelated current-main pyrefly error in torch/nn/modules/loss.py.

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -3791,6 +3797,8 @@ def call_function(
 class HintsWrapperHigherOrderVariable(WrapHigherOrderVariable):
     _HOP_NAME = "torch.ops.higher_order.hints_wrapper"
     _ALLOW_FALLBACK_TO_EAGER = False
+    supports_input_mutation = False
+    supports_aliasing = False
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py]: why do we need to set these to False if that's the default?

Comment (zou3519) [reply torch/_dynamo/variables/higher_order_ops.py]: ditto for a bunch of the places in this PR. that looked wrong to me, unless the goal is to be explicit

Comment (jansel) [reply torch/_dynamo/variables/higher_order_ops.py]: My agent says I removed the redundant explicit False settings where the new speculate_subgraph defaults already apply. The remaining explicit False override is only on hints_wrapper because it inherits WrapHigherOrderVariable, whose class defaults are permissive, and I added a short comment there to make that distinction clear.

**Review by aorenste — APPROVED** [review]

**Comment by jansel** [comment]
@zou3519 should take another look at this since he has prior comments

**Comment by jansel** [comment]
@zou3519 ping again for review

**Comment by jansel** [comment]
@zou3519 ping again for review

**Review by zou3519 — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

**Comment by jansel** [comment]
My agent says Dr.CI’s flex-attention failure came from mask_fn treating two read-only captured views of the same visibility tensor as unsupported aliasing. I split input-to-input support from output aliasing and enabled only the narrow mask_fn capability in forward and backward; input mutation and input/output alias checks remain strict. The exact failing CUDA test, focused and dynamic HOP tests, related suites, and lintrunner -a pass locally.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the August 12 review: the alias/mutation checker booleans are now keyword-only; register_hook explicitly rejects input mutation with fullgraph and graph-break fallback coverage; flex mask input-input aliasing has direct backward-HOP coverage; and WrapHigherOrderVariable is now a strict shared base with only concrete transparent wrappers opting in. The focused HOP/dynamic-shape/hooks/checkpoint/export suites, the prior CUDA CI reproducer, and lintrunner -a pass. I also verified that the public flex_attention path uses an internal fullgraph wrapper, so rejected mask output aliases surface as an error rather than silently falling back.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the three Dr.CI failures came from the new direct flex-attention backward test computing its eager reference inside the compiled-autograd harness, which nested compiled autograd under make_fx. The test is now CPU-capable and disables compiled autograd only for that eager reference while retaining compiled-path coverage. I also addressed the latest review by adding root strict defaults, forward/backward mutation coverage, non-fullgraph flex fallback coverage, named register_hook diagnostics, and a safe graph break for the previously mis-lowered sourceless post-accumulate hook path. Focused tests and lintrunner -a pass.

**Comment by jansel** [comment]
@claude review these changes
